### PR TITLE
Add charuco pattern into calibration.cpp

### PR DESCRIPTION
**Pull request by vovka643**
added charuco pattern into calibration.cpp. Added charuco pattern with predefined aruco dictionary and with dictionary from file.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [х] I agree to contribute to the project under Apache 2 License.
- [х] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [х] The PR is proposed to the proper branch
- [х] There is a reference to the original bug report and related work
- [х] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [х] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/cpp/calibration.cpp [hunk 1]
```diff
@@ -50,13 +52,16 @@ static void help(char** argv)
...
-        "     [-pt=<pattern>]          # the type of pattern: chessboard or circles' grid\n"
+        "     [-pt=<pattern>]          # the type of pattern: chessboard or charucoboard or circles' grid\n"
         "     [-n=<number_of_frames>]  # the number of frames to use for calibration\n"
         "                              # (if not specified, it will be set to the number\n"
         "                              #  of board views actually available)\n"
         "     [-d=<delay>]             # a minimum delay in ms between subsequent attempts to capture a next view\n"
         "                              # (used only for video capturing)\n"
         "     [-s=<squareSize>]        # square size in some user-defined units (1 by default)\n"
+        "     [-ms=<markerSize>]       # marker size in some user-defined units (0.5 by default)\n"
+        "     [-ad=<arucoDict>]        # number of predefined aruco dictionary"
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:63]: "Aruco dictionary name for charuco board"

File: samples/cpp/calibration.cpp [hunk 2]
```diff
@@ -50,13 +52,16 @@ static void help(char** argv)
...
+        "     [-adf=<dictFilename>]   # filename of aruco dictionary"
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:64]: "Custom aruco dictionary file for charuco board"

File: samples/cpp/calibration.cpp [hunk 3]
```diff
@@ -529,6 +543,40 @@ int main( int argc, char** argv )
...
+                if (dictFilename == "None") {
+                    dictionary = cv::aruco::getPredefinedDictionary(arucoDict);
+                }
+                else {
+                    cv::FileStorage dict_file(dictFilename, cv::FileStorage::Mode::FORMAT_JSON && cv::FileStorage::Mode::READ);
+                    cv::FileNode fn(dict_file.root());
+                    dictionary.readDictionary(fn);
+                }
+                cv::Ptr <cv::aruco::CharucoBoard> board =
+                    cv::makePtr<cv::aruco::CharucoBoard>(cv::aruco::CharucoBoard({ boardSize.width + 1,boardSize.height + 1 },
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:558]: No need to use `cv::Ptr` for board. You can allocate it on stack as local variable.

File: samples/cpp/calibration.cpp [hunk 4]
```diff
@@ -529,6 +543,40 @@ int main( int argc, char** argv )
...
+                // if at least one marker detected
+                if (markerIds.size() > 0) {
+                    std::vector<cv::Point2f> charucoCorners;
+                    std::vector<int> charucoIds;
+                    found = cv::aruco::interpolateCornersCharuco(markerCorners, markerIds, view, board, pointbuf, charucoIds);
+                    
+                    if (pointbuf.size() < boardSize.height*boardSize.width) {
+                        found = false;
+                    }
+                }
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:574]: No need to create yet another detector. CharucoBoard has `detectBoard` method. It does all you need.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: samples/cpp/calibration.cpp [hunk 5]
```diff
@@ -529,6 +542,42 @@ int main( int argc, char** argv )
...
+                if (dictFilename == "None") {
+                    dictionary = cv::aruco::getPredefinedDictionary(arucoDict);
+                }
+                else {
+                    cv::FileStorage dict_file(dictFilename, cv::FileStorage::Mode::FORMAT_JSON && cv::FileStorage::Mode::READ);
+                    cv::FileNode fn(dict_file.root());
+                    dictionary.readDictionary(fn);
+                }
+
+                cv::aruco::CharucoBoard ch_board ({ boardSize.width + 1, boardSize.height + 1 },
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:557]: Why +1?

File: samples/cpp/calibration.cpp [hunk 6]
```diff
@@ -529,6 +542,42 @@ int main( int argc, char** argv )
...
+                    squareSize, markerSize, dictionary);
+
+                std::vector<int> markerIds;
+                std::vector<std::vector<cv::Point2f>> markerCorners;
+                
+                cv::aruco::ArucoDetector aruco_detector(dictionary);
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:563]: ArucoDetector is not used.

File: samples/cpp/calibration.cpp [hunk 7]
```diff
@@ -529,6 +542,42 @@ int main( int argc, char** argv )
...
+                if (markerIds.size() > 0) {
+                    if (pointbuf.size() < boardSize.height*boardSize.width) {
+                        found = false;
+                    }
+                    else {
+                        found = true;
+                    }
+                }
+                else {
+                    found = false;
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:577]: You can initialize found as false beforehand and get rid of all else branches.

File: samples/cpp/calibration.cpp (hunk 5 above)
Comment (asmorkalov) [reply samples/cpp/calibration.cpp]: Still relevant.

**Comment by asmorkalov** [comment]
```
/home/ci/opencv/samples/cpp/calibration.cpp:552:122: warning: enum constant in boolean context [-Wint-in-bool-context]

/home/ci/opencv/samples/cpp/calibration.cpp:565:37: warning: comparison of integer expressions of different signedness: 'std::vector<cv::Point_<float> >::size_type' {aka 'long unsigned int'} and 'int' [-Wsign-compare]
```

**Comment by asmorkalov** [comment]
```
./example_cpp_calibration -w=7 -h=5 -pt=charucoboard -ad=DICT_5X5_100 ./calib.list -o=charuco_calib.yaml
...
ERRORS:
Parameter 'ad': can not convert: [DICT_5X5_100] to [int]
```

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by asmorkalov** [comment]
The sample uses wrong visualization or something is broken in the pipeline. I see this:
preview_1

**Comment by vovka643** [comment]
You should write inner corners of desk.  For your desk you should use parameters: h=4, w =6

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by asmorkalov** [comment]
Please ignore my previous statement about wrong visualization and calibration. I screwed up with board size. `./example_cpp_calibration -w=4 -h=6 -pt=chessboard ./calib.list -o=charuco_calib.yaml` works expected.

File: samples/cpp/calibration.cpp [hunk 8]
```diff
@@ -476,6 +476,20 @@ int main( int argc, char** argv )
...
+    cv::aruco::Dictionary dictionary;
+    if (dictFilename == "None") {
+        dictionary = aruco::getPredefinedDictionary(aruco::PredefinedDictionaryType(arucoDict));
+    }
+    else {
+        cv::FileStorage dict_file(dictFilename, cv::FileStorage::Mode::READ);
+        cv::FileNode fn(dict_file.root());
+        dictionary.readDictionary(fn);
+    }
+    cv::aruco::CharucoBoard ch_board({ boardSize.width + 1, boardSize.height + 1 },
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:521]: Why +1?

File: samples/cpp/calibration.cpp [hunk 9]
```diff
@@ -529,6 +589,17 @@ int main( int argc, char** argv )
             case ASYMMETRIC_CIRCLES_GRID:
                 found = findCirclesGrid( view, boardSize, pointbuf, CALIB_CB_ASYMMETRIC_GRID );
                 break;
+            case CHARUCOBOARD:
+            {
+                ch_detector.detectBoard(view, pointbuf, markerIds);
+                if (pointbuf.size() < (long int) (boardSize.height*boardSize.width)) {
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:595]: I propose to cast to size_t.

File: samples/cpp/calibration.cpp (hunk 8 above)
Comment (asmorkalov) [reply samples/cpp/calibration.cpp:521]: Discussed the issue offline. Decided to remove +1, but add note about the board size units. Chessboard and radon board size are defined in inner corners, all other boards - in elements (circles or squares).

File: samples/cpp/calibration.cpp (hunk 5 above)
Comment (asmorkalov) [reply samples/cpp/calibration.cpp]: Discussed the issue offline. Decided to remove +1, but add note about the board size units. Chessboard and radon board size are defined in inner corners, all other boards - in elements (circles or squares).

File: samples/cpp/calibration.cpp [hunk 10]
```diff
@@ -50,13 +51,22 @@ static void help(char** argv)
...
         "     [-d=<delay>]             # a minimum delay in ms between subsequent attempts to capture a next view\n"
         "                              # (used only for video capturing)\n"
         "     [-s=<squareSize>]        # square size in some user-defined units (1 by default)\n"
+        "     [-ms=<markerSize>]       # marker size in some user-defined units (0.5 by default)\n"
+        "     [-ad=<arucoDict>]        # Aruco dictionary name for charuco board\n"
+        "Available ArUco dictionaries: DICT_4X4_50, DICT_4X4_100, DICT_4X4_250, "
+        "DICT_4X4_1000, DICT_5X5_50, DICT_5X5_100, DICT_5X5_250, DICT_5X5_1000, "
+        "DICT_6X6_50, DICT_6X6_100, DICT_6X6_250, DICT_6X6_1000, DICT_7X7_50,"
+        "DICT_7X7_100, DICT_7X7_250, DICT_7X7_1000, DICT_ARUCO_ORIGINAL"
+        "DICT_APRILTAG_16h5, DICT_APRILTAG_25h9, DICT_APRILTAG_36h10, "
```
Comment (asmorkalov) [inline samples/cpp/calibration.cpp:67]: No space at the end of the line.

**Review by asmorkalov — APPROVED** [review]
:+1: Tested manually with command `./example_cpp_calibration -w=5 -h=7 -pt=charuco -ad=DICT_5X5_100 ./calib.xml -o=charuco_calib.yml` and template from https://github.com/opencv/opencv/pull/23363/

**Pull request merged**